### DISCUSSION TITLE
Build Error MingW 18  GNU 11.2.0 With Code blocks. Generated with Cmake 3.27.8

### DISCUSSION THREAD

**Discussion by IndieDev99** [vsg-dev/VulkanSceneGraph#1043] [Q&A] [answered]
Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};

I'm new and beginner so, if you guys can help me. My device is Quadro K620 vulkan 1.3 supported GPU. 16GB RAM Hexa Core CPU. Using Windows 10 pro 21H2. 1080p 60z monitor. Vulkan SDK installed 1.2.189.0.

I found something in code
  if (screenNum >= displayDevices.size()) throw Exception{"Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};

**Comment by robertosfield** [comment]
I don't use Windows myself byt my best guess is your have a driver issue an the Vulkan driver is reporting that there are no available devices to use.

Could you run vsgdeviceselection example found in vsgExamples:

~~~ sh
vsgdeviceselection --list
~~~

This is what I see on my Kubuntu 22.04 system that currently has a Geforce 2060 installed:

~~~ sh
$ vsgdeviceselection --list
vkEnumerateInstanceVersion() 4206860
VK_API_VERSION = 1.3.268.0

physicalDevices.size() = 2
    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7f85830bbc40) NVIDIA GeForce RTX 2060, deviceType = 2, apiVersion = 1.3.242, driverVersion = 23.516.192
    QueueFamilyProperties 4
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER | PARSE_BINDING, queueCount = 16, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
        VkQueueFamilyProperties[1] queueFlags = TRANSFER | PARSE_BINDING, queueCount = 2, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
        VkQueueFamilyProperties[2] queueFlags = COMPUTE | TRANSFER | PARSE_BINDING, queueCount = 8, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
        VkQueueFamilyProperties[3] queueFlags = TRANSFER | PARSE_BINDING | VIDEO_DECODE, queueCount = 1, timestampValidBits = 32, minImageTransferGranularity = {1, 1, 1}

    matched ref_ptr<vsg::PhysicalDevice>(vsg::PhysicalDevice 0x7f85830bc0b0) llvmpipe (LLVM 15.0.7, 256 bits), deviceType = 4, apiVersion = 1.3.238, driverVersion = 0.0.1
    QueueFamilyProperties 1
        VkQueueFamilyProperties[0] queueFlags = GRAPHICS | COMPUTE | TRANSFER, queueCount = 1, timestampValidBits = 64, minImageTransferGranularity = {1, 1, 1}
~~~

**Reply by IndieDev99** [reply]
Thanks !!
I'm trying

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/48a13e7b-e0c7-404d-8353-589234c2956b)
This is what i get after executing the vulkaninfo command in cmd.

**Reply by robertosfield** [reply]
Are you able to run vsgdeviceselection?

Are you able to run any other Vulkan applications?

Does the VSG build cleanly?

**Reply by IndieDev99** [reply]
Yes I am Running Doom 2016 Using Vulkan. Godot Engine Vulkan Demo and Unity Engine using Vulkan Renderer.

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/69ee6471-6179-4d71-93a7-e11cd52d966b)
Godot Demo

**Reply by IndieDev99** [reply]
Now I am Trying with Visual Studio 2019 for build. Lets see if it works.

**Reply by IndieDev99** [reply]
failed again in VS 2019. 
I will rebuild it again.

**Reply by robertosfield** [reply]
We haven't seen reports of problems like this from other Windows users so there much be something specific about your build/drivers/hardware that is causing problems.

When you say VS 2019 failed, what was the problem this time?

There far more recent VulkanSDK versions than the 1.2.189.0 - the latest is 1.3.268.0, perhaps it would be worth try updating:

[https://vulkan.lunarg.com/sdk/home](https://vulkan.lunarg.com/sdk/home)

**Reply by IndieDev99** [reply]
There is some other issue. I have done upgrading SDK to 1.3. And I have rebuild many times with different versions of VSG.

**Reply by robertosfield** [reply]
What issues are you coming across?  

Others are using Windows without problems so there much be something specific to what you are doing or how your system is setup, so there is something specific in your case that is going wrong.  The more info you can provide us the better we'll be able to help.

Have you run vsgdeviceselection from vsgExamples?

Have you been able to build any other 3rd party Vulkan applications?

**Reply by IndieDev99** [reply]
Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};

**Reply by IndieDev99** [reply]
Other Applications are working. Vulkan Cube is also working. Vulkan Commands are working.

**Reply by sbrkopac** [reply]
I just tried on Visual Studio 2022 and everything works correctly.

**Reply by IndieDev99** [reply]
How ?
Steps and Your setup ?
What I have done:
1)Installed Vulkan SDK 1.2 and 1.3
2)Latest Nvidia Drivers
3)Cmake
4)MingW 18 GNU 11.2.0
5) Using Cmake and GNU GCC with Codeblocks 20.03 I started Building the VSG

**Reply by robertosfield** [reply]
Could you post your CMakeCache.txt file?  Perhaps this might gives us some clues.

What application I you running when you get the "Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range."?

The code that throws this vsg::Exception is found in this part of the Win32_Window constructor:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/platform/win32/Win32_Window.cpp#L350

This exception is thrown when traits->screenNum >= displayDevices.size(), screenNum defaults to 0, so I'd expect it to be 0, so for this conditions to happen the displayDevices is most likely empty().  

Which suggests EnumDisplayDevices(nullptr, deviceNum, &displayDevice, 0) is not returning any usable devices.  This is a Win32 function call and not something the VSG has much input on.

My best guess at this point is that your install of MinGW doesn't pick up on Win32 correctly for some reason.

I presume that other applications you are trying are prebuilt in other environments so unaffected whatever is going on, on your machine.

**Reply by robertosfield** [reply]
While I'm not a Windows users myself, others have reported success with MinGW and Cigwin, but perhaps these are different versions/configurations.  Any chance that there might be some 32bit vs 64bit issue going on?

**Reply by IndieDev99** [reply]
@robertosfield let me try visual studio again. I will post the error.

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/bbe69fe2-ccbc-4859-80cc-32d5fc3b4caa)
code test for vulkan using the codeblocks 20.03 with mingw

**Reply by IndieDev99** [reply]
Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\Ms2019Build\CMakeFiles\c0c82d1a4ebb7110b6b25dc960e99185\uninstall-vsg.rule
Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\CMakeLists.txt' exited with code 1.	uninstall-vsg	C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets	241

**Comment by IndieDev99** [comment]
>
[CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/13515342/CMakeCache.txt)
 at there might be some 32bit vs 64bit issue going on

**Reply by robertosfield** [reply]
Are you working on 32 or 64bit?

**Reply by IndieDev99** [reply]
I don't understand. 
I'm building the VSG from source code. And I don't see any 32 or 64 bit like of anything. 
I have 64 bit Windows and all tools are 64 bit. don't know about VSG

**Reply by robertosfield** [reply]
Is MinGW 32 or 64bit?

The VSG is mostly agnostic about 32 or 64bit, the only place where there can be specific paths are where OS functions are used for Windows/file system.  Where possible we write agnostic code but with the OS level stuff we have to work with what is available.  

I wouldn't have though this would affect the EnumDisplayDevices(..) call but I'm not Windows user.

The best thing I can suggest at this point is to add some debugging into Win32_Window constructor to report the values returned from EnumDisplayDevices(), or stop the debugger around this code so you can inspect what values it's reporting.

FYI, you are the first person to report this specific issue so there is something odd going on, on your system.

**Reply by IndieDev99** [reply]
I'm using minGW 64 bit. 
And I will test it again. Let's see what happens next.

**Reply by robertosfield** [reply]
@IndieDev99 have you made any progress?  Were you able to get any details about what EnumDisplayDevices(..) is returning?

As it looks like the Win32_Window.cpp is failing the first hurdle of getting a device to open a window on and other Vulkan applications are working fine, perhaps there is something extra required to work on some Windows machines.  I say some Windows machines as your are the first Windows users to report this particular issue.  

Perhaps another avenue to look at would be trying to build other Vulkan examples on your machine, such as Sasha Willems:

https://github.com/SaschaWillems/Vulkan

These examples use glfw for Windowing so if these work and the VSG doesn't then comparing the glfw Win32 code to the VSG's might be helpful in spotting what might be different and missing from the VSG side.

**Reply by IndieDev99** [reply]
I am unable to build. Now using Godot but for C#. Let me test the Willems vulkan and see what happens.

**Reply by robertosfield** [reply]
What were the build issues?

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/2b1af946-f5fc-46a7-ad1f-e3accb57a20f)
https://github.com/SaschaWillems/Vulkan. Done Building

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/8b897920-7cfd-40ec-b037-1e4922062242)
Working. Vulkan Examples

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/47e611a3-810c-459f-a5eb-0ebc309898dc)

**Reply by IndieDev99** [reply]
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/124052807/0cf32fe7-5e9c-4fb2-a068-2f240919096b)
Every other vulkan app is working. Also Tried Vulkan with Ogre3d.  I'm thinking to use Godot for my games. May be I will use Ogre3d for C++. Thanks for your time and your support.

**Reply by robertosfield** [reply]
Good to see you've been able to get the Vulkan examples working fine.

This should it should be possible.  

You mentioned build issues but not specified what they are, could you point provide some details so we can work to address them?

In the initial build success but runtime failure we which suggests something going on with EnumDisplayDevices for some reason, as others haven't reported issues with this unfortunately reproducing it could be hard.  I can only presume that there is some combination of system setup that is unusually enough that it's tripping up what the VSG is trying to do to get the details of the monitors on your system.  

Is there any chance you are working within a virtual machine?

**Reply by IndieDev99** [reply]
No I'm not working in any virtual machine. It don't give any details when the error comes after the platform win32 build step reaches. That's why I posted it here. I have posted CmakeCache file.
[CMakeCache.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/13536707/CMakeCache.txt)

**Reply by robertosfield** [reply]
I'm confused, the title you gave for this thread is build error, the first post to this thread you mention a runtime error:

 "Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};" 

Which suggests you did get things to build, but elsewhere you talk about build errors, but I can see anywhere where you've posted the actual details of build errors.  The CMakeCache.txt just provides details of the dependencies and options it's selected, so of a bit of use but isn't related to build errors.

Could you post the build errors you see with details of the compiler/build environment you are using when you see them?

**Reply by IndieDev99** [reply]
||=== Build: all in vsg (compiler: GNU GCC Compiler) ===|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\include\vsg\platform\win32\Win32_Window.h|191|warning: type qualifiers ignored on function return type [-Wignored-qualifiers]|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp||In constructor 'vsgWin32::Win32Surface::Win32Surface(vsg::Instance*, HWND)':|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|47|warning: unused variable 'result' [-Wunused-variable]|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp||In constructor 'vsgWin32::Win32_Window::Win32_Window(vsg::ref_ptr<vsg::WindowTraits>)':|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|364|warning: comparison of integer expressions of different signedness: 'int32_t' {aka 'int'} and 'std::vector<_DISPLAY_DEVICEA>::size_type' {aka 'long long unsigned int'} [-Wsign-compare]|
.\||Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp||In member function 'LRESULT vsgWin32::Win32_Window::handleWin32Messages(UINT, WPARAM, LPARAM)':|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|640|warning: declaration of 'event_time' shadows a previous local [-Wshadow]|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|533|note: shadowed declaration is here|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|645|warning: declaration of 'event_time' shadows a previous local [-Wshadow]|
D:\Devlopement\C++\Vulkan Scene Graph\VulkanSceneGraph-1.0.9\VulkanSceneGraph-1.0.9\src\vsg\platform\win32\Win32_Window.cpp|533|note: shadowed declaration is here|
||=== Build finished: 1 error(s), 5 warning(s) (11 minute(s), 21 second(s)) ===|

**Reply by IndieDev99** [reply]
I'm building the VSG from source using GNU GCC and Codebocks. All other vulkan based egnines and apps are building even C++ 17 based engines. 
I followed the same instructions given but I'm not getting rid of  **### Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE}**

**Reply by timoore** [reply]
This is finally becoming clear: you are complaining about a line in the compilation that mentions the ".||Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};|" message.  This is not a compilation error: it looks to me that it is only causing a warning on that line. You do seem to have a compilation error somewhere else. 

Search for the string "error:" in the compilation output.

**Reply by IndieDev99** [reply]
> This is finally becoming clear: you are complaining about a line in the compilation that mentions the ".||Error: vsg::Win32_Window::Win32_Window(...) failed to create Window, screenNum is out of range.", VK_ERROR_INVALID_EXTERNAL_HANDLE};|" message.  This is not a compilation error: it looks to me that it is only causing a warning on that line. You do seem to have a compilation error somewhere else. 
> 
> Search for the string "error:" in the compilation output.

This is an error. Due to some reasons VSG is telling that window creation is failed due to something I cannot find. But I will try it tomorrow

**Reply by timoore** [reply]
Window creation isn't failing; you haven't succeeded in compiling the program and are not close to running it. I think you are being fooled by seeing "Error:" in the warning messages from the compilation.

**Comment by robertosfield** [comment]
Thanks looks your compiler is throwing issues that other compilers aren't.
I am to pub right now. Will look into it tomorrow.

**Reply by IndieDev99** [reply]
Thanks !!

**Reply by robertosfield** [reply, marked as answer]
I had a look at your snippet of build warnings and believe Tim is correct, the snippet doesn't contain any actual compile errors, it's simply that one of the lines with a warning about comparing two different integer types just happens to have a runtime error message.

I can see that at the end of the snippet your have 1 build error and 5 warnings, but I can't spot anything other than the 5 warnings in the snippet, so my best guess is that you've run a build, one of the files has failed to compile but the compile has continued  and the warnings have caused the compile error to scroll off the screen or something similar - something that can easily happen with a multi0threaded build.

We have automated build testing on the major platforms, which includes WIndows but this is for VisualStudio, we also have gcc build in the mix but these are not for Windows.  We don't have any build errors or warnings these build combinations so I presume that the Windows specific files that are being built on your system are throwing up more warnings and an error.  

The warnings should be straight forward to fix, but the we'll need a copy of the build error before we can attempt to tackle it.

Would it be possible to run the build and redirect the output to a file to capture all the output? 

Something like:

    make &> build_output.txt

**Reply by IndieDev99** [reply]
Thanks for looking into the problem. I am currently busy in a work. I will reinstall the Windows on my new PC. So, i think it will take months  before I can come back and start working with vsg again because I have found a job and I will be busy there. I will try this in my new pc. May be I will not comeback to vsg again if I continue with Godot.  If I get this problem solved I will let you know. Thanks again for your time !!

**Reply by robertosfield** [reply]
I have checked fixes for the reported warnings into MinGW_build_fixes branch:

https://github.com/vsg-dev/VulkanSceneGraph/compare/MinGW_build_fixes

If the automated builds run cleanly on Windows I'll merge this with VSG master.  They are only warning fixes though so won't resolve the build error, which alas will remain a mystery till someone can recreate it and post what the error is :-)

Good luck with the new job.

**Comment by robertosfield** [comment]
The warning fixes passed the Windows automated build so I've gone ahead and merged with VSG master with PR #1048.
